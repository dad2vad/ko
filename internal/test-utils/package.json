{
  "name": "@kosko/test-utils",
  "version": "0.0.0",
  "private": true,
  "engines": {
    "node": ">=14.18.0"
  },
  "main": "dist/index.node.cjs",
  "module": "dist/index.node.mjs",
  "types": "dist/index.d.ts",
  "sideEffects": false,
  "scripts": {
    "build": "build-scripts"
  },
  "exports": {
    "types": "./dist/index.d.ts",
    "deno": "./dist/index.deno.mjs",
    "node": {
      "import": "./dist/index.node.mjs",
      "require": "./dist/index.node.cjs"
    },
    "import": "./dist/index.base.mjs",
    "require": "./dist/index.node.cjs"
  },
  "dependencies": {
    "execa": "^5.1.1",
    "temp-dir": "^2.0.0",
    "tmp": "^0.2.1",
    "tmp-promise": "^3.0.3"
  },
  "devDependencies": {
    "@kosko/build-scripts": "workspace:^",
    "@types/tmp": "^0.2.3"
  }
}
