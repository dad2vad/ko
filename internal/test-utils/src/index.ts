export * from "./getRejectedValue";
export * from "./installPackage";
export * from "./makeTempDir";
export * from "./makeTempFile";
export * from "./projectRoot";
