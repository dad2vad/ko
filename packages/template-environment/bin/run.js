#!/usr/bin/env node

require("@kosko/template").run(require("../dist/index.node.cjs").default);
