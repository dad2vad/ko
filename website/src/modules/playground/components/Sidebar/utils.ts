export function isValidName(value: string): boolean {
  return /^[\w-_.]+$/.test(value);
}
