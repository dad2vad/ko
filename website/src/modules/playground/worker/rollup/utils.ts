export function getModuleURLForCDN(module: string) {
  return `https://jspm.dev/${module}`;
}
