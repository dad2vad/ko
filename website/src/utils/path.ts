export const sep = "/";

export function isAbsolute(path: string): boolean {
  return path.startsWith(sep);
}

export function isRelative(path: string): boolean {
  return path.startsWith(`.${sep}`) || path.startsWith(`../${sep}`);
}

export function dirname(path: string): string {
  const index = path.lastIndexOf(sep);
  if (index === -1) return ".";
  return path.substring(0, index) || sep;
}

export function extname(path: string): string {
  const index = path.lastIndexOf(".");
  if (index === -1) return "";
  return path.substring(index);
}

export function basename(path: string, ext?: string): string {
  const index = path.lastIndexOf(sep);
  const base = index === -1 ? path : path.substring(index + 1);

  if (base.endsWith(ext)) {
    return base.substring(0, base.length - ext.length);
  }

  return base;
}

export function relative(from: string, to: string): string {
  const fromParts = from.split(sep).filter(Boolean);
  const toParts = to.split(sep).filter(Boolean);

  while (fromParts[0] && toParts[0]) {
    fromParts.shift();
    toParts.shift();
  }

  while (toParts[0] === "." || toParts[0] === "..") {
    const toPart = toParts.shift();

    if (toPart === "..") {
      fromParts.pop();
    }
  }

  while (fromParts.pop()) {
    toParts.unshift("..");
  }

  return toParts.join(sep);
}
