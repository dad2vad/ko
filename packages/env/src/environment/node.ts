import { getRequireExtensions } from "@kosko/require";
import { join } from "node:path";
import { formatPath } from "../paths";
import { toArray } from "@kosko/common-utils";
import {
  createEnvironment,
  createLoaderReducers,
  createReducerList
} from "./base";
import { Environment, Loader, ReducerExecutor, ReducerList } from "./types";
import { cwd as getCwd } from "node:process";

export function getNodeExtensions(): string[] {
  return getRequireExtensions().map((ext) => ext.substring(1));
}

/**
 * Options for creating a Node.js environment.
 *
 * @public
 */
export interface NodeEnvironmentOptions {
  /**
   * Current working directory (CWD).
   *
   * @defaultValue `process.cwd()`
   */
  cwd?: string;
}

export interface InternalNodeEnvironmentOptions extends NodeEnvironmentOptions {
  createReducerExecutor(this: void, reducers: ReducerList): ReducerExecutor;
  requireModule(this: void, environment: Environment, id: string): any;
  mergeValues(this: void, data: any[]): any;
}

export function createNodeEnvironment({
  cwd = getCwd(),
  createReducerExecutor,
  requireModule,
  mergeValues
}: InternalNodeEnvironmentOptions): Environment {
  function requireAllEnvs(template: string, component?: string) {
    if (!environment.env) return [];

    const envs = toArray(environment.env);

    return envs.map((env) => {
      const path = formatPath(template, {
        environment: env,
        ...(component && { component })
      });

      return requireModule(environment, join(environment.cwd, path));
    });
  }

  const loader: Loader = {
    global: () => requireAllEnvs(environment.paths.global),
    component: (name) => requireAllEnvs(environment.paths.component, name)
  };
  const reducers = createReducerList(createLoaderReducers(loader, mergeValues));
  const { reduce } = createReducerExecutor(reducers);
  const environment = createEnvironment({
    cwd,
    extensions: getNodeExtensions(),
    setReducers: reducers.setReducers,
    resetReducers: reducers.resetReducers,
    reduce
  });

  return environment;
}
