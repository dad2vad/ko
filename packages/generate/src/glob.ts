import { readdir } from "node:fs/promises";
import { join, posix } from "node:path";
import mm from "micromatch";

function createMatcher(patterns: readonly string[], baseOptions: mm.Options) {
  const positivePatterns: string[] = [];
  const negativePatterns: string[] = [];

  for (const pattern of patterns) {
    const parsed = mm.parse(pattern, baseOptions) as [
      { negated: boolean; consumed: string }
    ];

    if (parsed.length && parsed[0].negated) {
      negativePatterns.push(parsed[0].consumed);
    } else {
      positivePatterns.push(pattern);
    }
  }

  const options: mm.Options = {
    ...baseOptions,
    ignore: negativePatterns
  };

  return (path: string): boolean => {
    return mm.isMatch(path, positivePatterns, options);
  };
}

async function readDir(path: string) {
  const files = await readdir(path, { withFileTypes: true });

  return files.sort((a, b) => {
    if (a.name > b.name) {
      return 1;
    }

    if (a.name < b.name) {
      return -1;
    }

    return 0;
  });
}

export interface GlobOptions {
  path: string;
  patterns: readonly string[];
  extensions: readonly string[];
}

export interface GlobResult {
  relativePath: string;
  absolutePath: string;
}

export async function* glob(options: GlobOptions): AsyncIterable<GlobResult> {
  const matcherOptions: mm.Options = { cwd: options.path };
  const suffix =
    options.extensions.length > 1
      ? `.{${options.extensions.join(",")}}`
      : `.${options.extensions.join(",")}`;
  const dirMatcher = createMatcher(options.patterns, matcherOptions);
  const fileMatcher = createMatcher(
    options.patterns.map((pattern) => pattern + suffix),
    matcherOptions
  );

  async function* walk(
    path: string,
    parent: string
  ): AsyncIterable<GlobResult> {
    for (const file of await readDir(path)) {
      const result: GlobResult = {
        relativePath: posix.join(parent, file.name),
        absolutePath: join(path, file.name)
      };

      if (file.isDirectory()) {
        if (dirMatcher(result.relativePath)) {
          yield result;
        }

        yield* walk(result.absolutePath, result.relativePath);
      } else if (fileMatcher(result.relativePath)) {
        yield result;
      }
    }
  }

  yield* walk(options.path, "");
}
