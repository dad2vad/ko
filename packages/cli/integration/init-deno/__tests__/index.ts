import { makeTempDir, TempDir } from "@kosko/test-utils";
import glob from "fast-glob";
import { join, posix } from "node:path";
import { readFile } from "node:fs/promises";
import { runDenoCLI } from "../../utils/run";
import { ExecaReturnValue } from "execa";
import replaceString from "replace-string";

let tmpDir: TempDir;
let result: ExecaReturnValue;

beforeEach(async () => {
  tmpDir = await makeTempDir();
  result = await runDenoCLI(["init", tmpDir.path]);
});

afterEach(async () => {
  await tmpDir.cleanup();
});

test("should setup a new folder", async () => {
  const paths = await glob("**/*", { cwd: tmpDir.path });
  const files: Record<string, string> = {};

  for (const path of paths) {
    files[posix.normalize(path)] = await readFile(
      join(tmpDir.path, path),
      "utf8"
    );
  }

  expect(files).toMatchSnapshot();
});

test("should print usage guide to console", () => {
  expect(
    replaceString(result.stderr, tmpDir.path, "<INIT_PATH>")
  ).toMatchSnapshot();
});
