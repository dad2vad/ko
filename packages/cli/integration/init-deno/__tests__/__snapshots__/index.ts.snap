// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should print usage guide to console 1`] = `
"[32minfo [39m - Creating a Kosko project in "<INIT_PATH>"
[32minfo [39m - Project created at "<INIT_PATH>"

Inside that directory, you can run several commands:

  [36mdeno task kosko generate[39m
    Validate components and generate Kubernetes manifests.

  [36mdeno task kosko validate[39m
    Only validate components.

We suggest that you begin by typing:

  [36mcd <INIT_PATH>[39m
  [36mdeno task kosko generate[39m"
`;

exports[`should setup a new folder 1`] = `
{
  "README.md": "This is a [Kosko] project bootstrapped with [\`kosko init\`][kosko] command.

## Getting Started

Generate Kubernetes manifests with the following command.

\`\`\`sh
deno task kosko generate
\`\`\`

The result will be printed in the console. You can pipe the output to \`kubectl\` to apply generated manifests on Kubernetes.

\`\`\`sh
deno task kosko generate | kubectl apply -f -
\`\`\`

When \`--env\` option is set, Kosko will read variables from environment files. Then you can fetch variables with \`@kosko/env\` package.

\`\`\`sh
deno task kosko generate --env dev
\`\`\`

You can try editing files in \`components\` and \`environments\` folder and re-run \`npm run generate\` to see changes.

## Learn More

Check [Kosko docs](https://kosko.dev/docs/) for more information about Kosko.

[kosko]: https://kosko.dev/
",
  "components/nginx.ts": "import { Deployment } from "kubernetes-models/apps/v1/Deployment";
import { Service } from "kubernetes-models/v1/Service";
import env from "@kosko/env";

const params = await env.component("nginx");

const name = "nginx";
const labels = { app: name };
const port = 80;

const deployment = new Deployment({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    replicas: params.replicas,
    selector: {
      matchLabels: labels
    },
    template: {
      metadata: { labels },
      spec: {
        containers: [
          {
            name: "nginx",
            image: "nginx:stable",
            ports: [{ containerPort: port }],
            env: [{ name: "NGINX_PORT", value: \`\${port}\` }]
          }
        ]
      }
    }
  }
});

const service = new Service({
  metadata: {
    name,
    namespace: params.namespace
  },
  spec: {
    selector: labels,
    ports: [{ port }]
  }
});

export default [deployment, service];
",
  "deno.json": "{
  "importMap": "./import_map.json",
  "tasks": {
    "kosko": "deno run -A npm:kosko"
  }
}
",
  "env.ts": "import env, { Environment as BaseEnvironment } from "npm:@kosko/env";

// Declare types for global environment variables
export interface GlobalEnvironment {
  namespace: string;
}

// Declare types for component environment variables
export interface ComponentEnvironment {
  // Fallback type of all other component variables which are not specified below
  [key: string]: unknown;

  // Specify types for each component
  nginx: {
    replicas: number;
  };
}

// Extend Environment interface
export interface Environment extends BaseEnvironment {
  global(): Promise<GlobalEnvironment>;

  component<K extends string>(
    name: K
  ): Promise<GlobalEnvironment & ComponentEnvironment[K]>;
}

export default env as Environment;
",
  "environments/dev/index.ts": "import { GlobalEnvironment } from "@kosko/env";

const env: GlobalEnvironment = {
  namespace: "dev"
};

export default env;
",
  "environments/dev/nginx.ts": "import { ComponentEnvironment } from "@kosko/env";

const env: ComponentEnvironment["nginx"] = {
  replicas: 3
};

export default env;
",
  "import_map.json": "{
  "imports": {
    "@kosko/env": "./env.ts",
    "kubernetes-models/": "https://esm.sh/kubernetes-models/"
  }
}
",
  "kosko.toml": "components = ["*"]
",
}
`;
